code/modules/crafting/craft.dm

```diff
@@ -248,6 +249,12 @@
 		else
 			for(var/i in 1 to needed_amount)
 				var/atom/movable/part_atom = locate(thing) in (surroundings - parts_used)
+				for(var/j in 1 to length(surroundings - parts_used))
+					if(!recipe.blacklist.Find(part_atom.type))
+						break
+					blacklisted += part_atom
+					part_atom = locate(thing) in (surroundings - parts_used - blacklisted)
```

Comment:
```suggestion
				var/atom/movable/part_atom
				for(var/atom/movable/candidate as anything in (surroundings - parts_used))
					if(!is_type_in_list(candidate, recipe.blacklist)
						part_atom = candidate
						break
```
It took me a while to understand what the loop was doing.
The suggested way will do the following:
Loop through all surrounding atoms minus the parts that are used. Break when it has found an atom which is not part of the recipe's blacklist. Then use that atom as the part.

This way you don't have to do a lookup each loop to find an appropriate atom to check. Also, this ensures that the last atom checked won't get used regardless of type.